src/getindex.jl

```diff
@@ -0,0 +1,143 @@
+const Indexer = AbstractVector{<:Integer}
+
+Base.IndexStyle(::LinearMap) = IndexCartesian()
+# required in Base.to_indices for [:]-indexing
+Base.eachindex(::IndexLinear, A::LinearMap) = Base.OneTo(length(A))
+Base.lastindex(A::LinearMap) = last(eachindex(IndexLinear(), A))
+Base.firstindex(A::LinearMap) = first(eachindex(IndexLinear(), A))
+
+function Base.checkbounds(A::LinearMap, i, j)
+    Base.checkbounds_indices(Bool, axes(A), (i, j)) || throw(BoundsError(A, (i, j)))
+    nothing
+end
+# Linear indexing is explicitly allowed when there is only one (non-cartesian) index
+function Base.checkbounds(A::LinearMap, i)
+    Base.checkindex(Bool, Base.OneTo(length(A)), i) || throw(BoundsError(A, i))
+    nothing
+end
+# checkbounds in indexing via CartesianIndex
+Base.checkbounds(A::LinearMap, i::Union{CartesianIndex{2}, AbstractVecOrMat{CartesianIndex{2}}}) =
+    Base.checkbounds_indices(Bool, axes(A), (i,)) || throw(BoundsError(A, i))
+Base.checkbounds(A::LinearMap, I::AbstractMatrix{Bool}) =
+    axes(A) == axes(I) || throw(BoundsError(A, I))
+
+# main entry point
+function Base.getindex(A::LinearMap, I...)
+    # TODO: introduce some sort of switch?
+    @boundscheck checkbounds(A, I...)
+    _getindex(A, Base.to_indices(A, I)...)
+end
+# quick pass forward
+Base.@propagate_inbounds Base.getindex(A::ScaledMap, I...) = A.λ .* A.lmap[I...]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, I...) = A.lmap[I...]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, i::Integer) = A.lmap[i]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, i::Integer, j::Integer) = A.lmap[i,j]
+
+########################
+# linear indexing
+########################
+function _getindex(A::LinearMap, i::Integer)
+    i1, i2 = Base._ind2sub(axes(A), i)
+    return _getindex(A, i1, i2)
+end
+_getindex(A::LinearMap, I::Indexer) = [_getindex(A, i) for i in I]
+_getindex(A::LinearMap, ::Base.Slice) = vec(Matrix(A))
+_getindex(A::LinearMap, I::Vector{CartesianIndex{2}}) = [(@inbounds A[i]) for i in I]
+_getindex(A::LinearMap, I::Base.LogicalIndex) = [(@inbounds A[i]) for i in I]
+
+########################
+# Cartesian indexing
+########################
+_getindex(A::LinearMap, i::Union{Integer,Indexer}, j::Integer) = (@inbounds (A*unitvec(A, 2, j))[i])
+_getindex(A::LinearMap, ::Base.Slice, j::Integer) = A*unitvec(A, 2, j)
+function _getindex(A::LinearMap, i::Integer, J::Indexer)
+    # try
+        # requires adjoint action to be defined
```

Comment:
would be helpful here to have a `throw` advising the user to implement an adjoint action instead of just failing on the e'A multiply?  (perhaps there is already such a message in the multiply routine, but even so the user might be thinking they are slicing or indexing not multiplying so a throw message here could still be helpful?

Reply:
I'm fine with either way. Currently the usual "adjoint not defined for $(typeof(A))" error exception is thrown out of a multiply routine, as you say. Throwing an error out of the getindex routine is perhaps indeed more informative, so users can think about whether they want to "work out" the adjoint, or whether they want to avoid the slicing.